monitor/loadgen/xcall.go

```diff
@@ -0,0 +1,131 @@
+package loadgen
+
+import (
+	"context"
+	"fmt"
+	"math/rand/v2"
+	"time"
+
+	"github.com/omni-network/omni/contracts/bindings"
+	"github.com/omni-network/omni/lib/contracts"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/ethclient/ethbackend"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/lib/netconf"
+	"github.com/omni-network/omni/lib/xchain"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	dead = "0x000000000000000000000000000000000000dead"
+)
+
+type XCallConfig struct {
+	NetworkID   netconf.ID
+	XCallerAddr common.Address
+	Period      time.Duration
+	Backends    ethbackend.Backends
+	Chains      []netconf.Chain
+}
+
+func xCallForever(ctx context.Context, cfg *XCallConfig) {
+	log.Info(ctx, "Starting periodic xcalls", "period", cfg.Period)
+
+	nextPeriod := func() time.Duration {
+		jitter := time.Duration(float64(cfg.Period) * rand.Float64() * loadgenJitter) //nolint:gosec // Weak random ok for load tests.
+		return cfg.Period + jitter
+	}
+
+	timer := time.NewTimer(nextPeriod())
+	defer timer.Stop()
+
+	// returns random chain pair
+	getChainPair := func() (netconf.Chain, netconf.Chain) {
+		i := rand.IntN(len(cfg.Chains)) //nolint:gosec // Weak random ok for load tests.
+		j := rand.IntN(len(cfg.Chains)) //nolint:gosec // Weak random ok for load tests.
+		for i == j {
+			j = rand.IntN(len(cfg.Chains)) //nolint:gosec // Weak random ok for load tests.
+		}
+
+		return cfg.Chains[i], cfg.Chains[j]
+	}
+
+	// tick immediately
+	if err := xCall(ctx, cfg, getChainPair); err != nil {
+		log.Warn(ctx, "Failed to xcall (will retry)", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			if err := xCall(ctx, cfg, getChainPair); err != nil {
+				log.Warn(ctx, "Failed to xcall (will retry)", err)
+			}
+			timer.Reset(nextPeriod())
+		}
+	}
+}
+
+func xCall(ctx context.Context, cfg *XCallConfig, getChainPair func() (from netconf.Chain, to netconf.Chain)) error {
+	fromChain, dstChain := getChainPair()
```

Comment:
nit: we commonly use `srcChain`, not `fromChain`

Reply:
https://github.com/omni-network/omni/pull/2723/commits/d78c48c1ea9954e550af3cb65161f4a1e254df88